### PR TITLE
SAM L21 i2c support

### PR DESCRIPTION
**Pull request by ant9000**

### CODE REVIEW

**Comment by dylad** [comment]
@ant9000 I tested your PR on my saml21-xpro and an accelerometer (ADXL345)
It works so far but I got a problem with I2C speed. The measured frequency is about 190kHz using my logic analyzer.
Could you confirm this problem ?

**Comment by ant9000** [comment]
Can't confirm at present, but on my YARM board I use the PLL (logic was taken from samr21-xpro):

https://github.com/ant9000/RIOT/blob/master/boards/yarm/include/periph_conf.h#L32

Maybe the same stuff should apply to saml21-xpro?

Antonio

**Comment by dylad** [comment]
I tried using the **same** PLL config. Now i2c frequency seems to be around 33kHz.

**Comment by dylad** [comment]
GCLK->GENCTRL[0] is set to 16Mhz in cpu/saml21/cpu.c (div factor = 0) and this is the clock source of SERCOM2 (used by your I2C)
So if you set CLOCK_DIV to 1 in periph_conf, CLOCK_CORECLOCK will be 16MHz too.
I2C frequency is okay with these settings. (Tried both 100kHz and 400kHz speed).

Someone can confirm ?
Regards

**Comment by ant9000** [comment]
@dylad took a scope  and tested the frequency on my YARM board - with a divisor of 1 it is indeed correct, even if a little bit jittery. I wrongly assumed that the code was correct since the peripheral I've tested it with was working - thanks a lot for pointing this matter out.

And I also noticed that PLL defines weren't used at all in SAM L21 initialization... sorry, my fault, again!

**Comment by dylad** [comment]
Hi,
Does this PR needs to comply with the new I2C interface introduced by #6575 and #6576  before merging it ? Or should we need another PR ?
Regards

File: cpu/sam0_common/periph/i2c.c [hunk 1]
```diff
@@ -35,11 +35,21 @@
 
 #define SAMD21_I2C_TIMEOUT  (65535)
 
+#define CPU_MODEL_SAML21 (CPU_MODEL_SAML21E18A || CPU_MODEL_SAML21G18A || \
```
Comment (aabadie) [inline cpu/sam0_common/periph/i2c.c]: I think you don't need this define, you can use the predefined macro `CPU_FAM_SAML21` instead.

**Comment by aabadie** [comment]
This PR needs rebase.
@haukepetersen as stated above, do you think this PR complies with #6575 and #6576 ?

**Comment by dylad** [comment]
@ant9000 Are you still working on it ?
Any chances to see this PR merged for April release ?

**Comment by dylad** [comment]
ping @ant9000

**Comment by ant9000** [comment]
@dylad sorry for not responding for so long.

I am currently too busy elsewhere, and unable to keep focus on RIOT development. Can't say when I'll be able to dedicate a non trivial amount of time again - hopefully not too long in the future.

Antonio

**Comment by dylad** [comment]
@ant9000 thanks a lot for informing us :)
What should we do about this PR then ?

**Comment by ant9000** [comment]
The PR stayed there staling for more than a month - probably I was the only one looking for I2C support on this CPU.

Unluckily, I'm not anymore in the position to check if my code complies to your other PRs. The patch in itself is really small, if you find it useful you can just copy/paste the code wherever you please and close the PR.

Sorry for the time lost,

Antonio

**Comment by biboc** [comment]
I2C on SAML21 should work the same than on SAMR21, is there any difference? Otherwise use i2c from SAMR21

**Comment by dylad** [comment]
@biboc There are some differences about clock management between SAMD21 family and SAML21 family. We just need to handle these small differences. We could easily get a shared driver for both family.

**Comment by aabadie** [comment]
It would be great to move forward on this one. @ant9000 and @dylad do you have time to address comments and test this before end of next week ?
We don't have to wait for #6575 and #6576. Note that this PR needs rebase.

**Comment by dylad** [comment]
I can propose a PR to @ant9000's branch or open another one on RIOT depending on @ant9000's availability.
But I cannot propose anything before the end of the month. Best I can do is early July.

**Comment by aabadie** [comment]
@dylad, I allowed myself to directly update this PR with the comments applied. Not sure if it still works. Can you test again (I might have missed something) ?

**Comment by dylad** [comment]
Thanks @aabadie !
I'll test it next week and share results.

**Comment by dylad** [comment]
@aabadie I tested I2C driver with my ADXL345 module, it works perfectly !

**Pull request merged**